### PR TITLE
fix: add handle for resolveExternalUri

### LINKED ISSUE #1936: asExternalUri doesn't work
Hi Guys,

I had written a new extension which renders Graphs from matplotlib into vscode webview component, it works great on vscode desktop, but I encounter problems when trying to work with it in code-server.

I cannot state what exactly the problem is but here is the workflow:
1. my vscode extension starts and listens to requests for plot viewing on port 11111 (POST request)
2. matplotlib custom backend sends a request to the extension with the url to open in webview
3. I use vscode.env.asExternalUri tunneling in order to open the url in a webview and embed the generated address in an Iframe.
4. Usually by now the webview with the webpage would have been loaded, but instead I get and error openning localhost refuse to connect

Screen Shot 2020-08-02 at 11 50 37

Any ideas what went wrong?

- Web Browser: Chrome
- Local OS: Ubuntu
- Remote OS: MacOS
- Remote Architecture: x86_64
- `code-server --version`: v3.4.1

### PR DESCRIPTION
**Pull request by jsjoeio**
This PR fixes `asExternalUri` to use code-server's built-in proxy or `VSCODE_PROXY_URI` if it's been set. 

Todos
- [x] handle path-based proxies
- [x] add e2e test for it
- [x] add e2e test for `VSCODE_PROXY_URI`
- [x] support `VSCODE_PROXY_URI`

Fixes #1936

### CODE REVIEW

File: patches/proxy-uri.diff
Comment (jsjoeio) [inline patches/proxy-uri.diff:133]: Everything above this comment was not made by me but by other patches. I could move my changes into a new patch if we think that's better. Open to either!

File: patches/proxy-uri.diff
Comment (jsjoeio) [inline patches/proxy-uri.diff:138]: Wondering if there's a better way to determine the `baseUrl`. `relativeRoot` is something we use but it's not available on the browser side AFAIK.

File: patches/proxy-uri.diff
Comment (jsjoeio) [inline patches/proxy-uri.diff:143]: I tried `extractLocalHostUriMetaDataForPortMapping` but was having trouble getting it to work for all `localhost` situations. Might be able to revisit.

**Comment by jsjoeio** [comment]
> Hunk #2 FAILED at 486.
> Hunk #*** FAILED at 497.
> Hunk #4 succeeded at 508 (offset 1 line).
> 2 out of 4 hunks FAILED -- rejects in file lib/vscode/src/vs/code/browser/workbench/workbench.ts
> Patch patches/proxy-uri.diff does not apply (enforce with -f)

Hmm...wondering if I need to refresh other patches or something. Any ideas?

File: patches/proxy-uri.diff
Comment (code-asher) [reply patches/proxy-uri.diff:143]: Was it for the no port scenarios?  If it returns undefined for those IMO we just return unchanged.

Comment (jsjoeio) [reply patches/proxy-uri.diff:143]: Yes, exactly!

> If it returns undefined for those IMO we just return unchanged

That means `http://localhost` and `https://localhost` would not be transformed to use `/proxy/<80 | 443>`. Sounds like we're okay with that so I'll update 👍🏼

File: patches/proxy-uri.diff
Comment (code-asher) [reply patches/proxy-uri.diff:138]: We need to resolve against the path in case code-server is hosted at a sub-path which means we have to get `proxyEndpointTemplate` in here somehow as it contains the information needed to resolve relatively.  Then we can do:

```
resolved = new URL(proxyEndpointTemplate.replace('{port}', port), window.location.href)
```

File: patches/proxy-uri.diff
Comment (code-asher) [reply patches/proxy-uri.diff:143]: If we do need to modify it we could patch the original `extractLocalHostUriMetaDataForPortMapping`?

File: patches/proxy-uri.diff
Comment (code-asher) [inline patches/proxy-uri.diff:109]: This might have been left over from experimentation.  You are right it is not available on browser (you can only use things in `common` or `browser`).  It would not work anyway as it needs the server-side request.  This is actually why we have `proxyEndpointTemplate` to pass the relative root to the browser!

File: patches/proxy-uri.diff
Comment (code-asher) [reply patches/proxy-uri.diff:133]: It looks like this patch now contains part of `base-path.diff`?  Not completely sure.

File: patches/proxy-uri.diff
Comment (code-asher) [reply patches/proxy-uri.diff:143]: Yeah I think that is chill because it gives us parity with upstream!  :tada:

File: patches/proxy-uri.diff
Comment (jsjoeio) [reply patches/proxy-uri.diff]: Super helpful! Thank you!

File: patches/proxy-uri.diff
Comment (code-asher) [inline patches/proxy-uri.diff:127]: I think if the production configuration does not exist we should error because we have no way of knowing what the real base is.  In the enterprise product for example this will end up being the wrong URI.

**Comment by jsjoeio** [comment]
Oops..looks like `proxyEndpointTemplate` adds `/proxy/port` for you.

image

File: patches/proxy-uri.diff
Comment (code-asher) [inline patches/proxy-uri.diff:129]: We can skip the `URI.parse` because `proxyEndpointTemplate` already includes `/proxy/{{port}}`.  For example it could be something like `../../proxy/{{port}}` then the `new URL` line will transform this into `https://mydomain.com/some/base/path/proxy/8080`.

**Review by code-asher — COMMENTED** [review]
Awesome work!

File: patches/proxy-uri.diff
Comment (jsjoeio) [reply patches/proxy-uri.diff]: Ah got it. okay I'll throw an error here.

File: patches/proxy-uri.diff
Comment (jsjoeio) [reply patches/proxy-uri.diff]: Updated!

**Comment by jsjoeio** [comment]
Looks good now! (ignore expected, hard-coded)

image

**Review by code-asher — COMMENTED** [review]
Looks great to me!  Do you want an approval now and merge or are you going to add the domain-based proxy in this same PR so I should approve after that is done?

File: patches/proxy-uri.diff
Comment (code-asher) [reply patches/proxy-uri.diff]: Just realized we do need the `URI.parse` just not the inner part but you already figured that out.  My bad!

**Comment by jsjoeio** [comment]
> are you going to add the domain-based proxy in this same PR so I should approve after that is done?

I'll work on that here then mark PR ready when all is said and done. Thanks for the early review!

**Comment by jsjoeio** [comment]
It appears the display language e2e test is failing. I'm going to see if I can reproduce locally.

**Comment by ksylvan** [comment]
Would you like me to try testing this?

**Comment by jsjoeio** [comment]
@ksylvan that would be super helpful 🙏🏼

File: patches/service-worker.diff
Comment (jsjoeio) [inline patches/service-worker.diff:64]: Looks like this patch update was just moving this down in the `.diff` file.

File: test/e2e/extensions/ms-ceintl.vscode-language-pack-es-1.70.0/package.json
Comment (jsjoeio) [inline test/e2e/extensions/ms-ceintl.vscode-language-pack-es-1.70.0/package.json:37]: I noticed VS Code added this after running the test so I decided to commit it.

File: test/e2e/models/CodeServer.ts
Comment (jsjoeio) [inline test/e2e/models/CodeServer.ts:330]: gonna drop this change

File: test/playwright.config.ts
Comment (jsjoeio) [inline test/playwright.config.ts:11]: saves time from having to look this up each time 😂

**Comment by jsjoeio** [comment]
@code-asher ready for another review!

File: test/e2e/extensions/ms-ceintl.vscode-language-pack-es-1.70.0/package.json
Comment (code-asher) [inline test/e2e/extensions/ms-ceintl.vscode-language-pack-es-1.70.0/package.json:36]: Was this necessary to fix something?  Might be sketchy to include stuff that comes from the closed-source marketplace.

**Review by code-asher — APPROVED** [review]
Nicely done!  I only have a comment about the added metadata.

File: test/playwright.config.ts
Comment (code-asher) [reply test/playwright.config.ts:11]: 100% lol same reason I added the other comments I am always forgetting the flags

File: test/e2e/extensions/ms-ceintl.vscode-language-pack-es-1.70.0/package.json
Comment (code-asher) [reply test/e2e/extensions/ms-ceintl.vscode-language-pack-es-1.70.0/package.json:36]: Oh somehow missed your comment where you said VS Code added it.  In that case seems fine to me!

**Comment by ericzhucode** [comment]
I found there's a conflict with url displayed in open external uri dialog.
If you open an external uri for the first time then the domain of it will be resolved to cache in map data structure.
Like https://www.w3schools.com/c/c_structs.php
image

For the next time if you open a different uri with the same domain
Like https://github.com/coder/code-server/issues
it will be resolved as the last uri which has been cached.
image

The uri displayed here is not correct
Hope there's a solutaion for this. Thank you!

**Pull request merged**
